src/software/backend/simulator_backend.h

```diff
@@ -41,57 +48,41 @@ class SimulatorBackend : public Backend
+     * @param physics_time_step Time time step used by the physics simulation
+     * @param world_time_increment The increment in the timestamp of the published World
+     * object each time it is published. ie. from the perspective of consumers, the World
+     * is being published every 'world_time_increment' seconds.
+     * @param simulation_speed_mode What speed mode to run the simulator with
      */
-    explicit SimulatorBackend(
-        const Duration& simulation_time_step        = Duration::fromSeconds(1.0 / 60.0),
-        unsigned int num_steps_per_primitive_update = 2);
+    explicit SimulatorBackend(const Duration& physics_time_step,
+                              const Duration& world_time_increment,
+                              SimulationSpeed simulation_speed_mode);
 
     /**
-     * Returns the current state of the World in the simulated backend
+     * Sets the SimulationSpeed for the simulation
      *
-     * @return the current state of the World in the simulated backend
+     * @param simulation_speed_mode the new SimulationSpeed to set
      */
-    World getWorld();
+    void setSimulationSpeed(SimulationSpeed simulation_speed_mode);
 
     /**
-     * Sets the state of the World in the simulated backend. All following simulation will
-     * follow from this newly provided world.
+     * Simulates the world given the initial state until the simulation times out
      *
-     * @param new_world A new state of the World to set in the simulated backend
+     * @param world The initial state of the world in the simulation
+     * @param timeout How long to run the simulation for before failing
+     * @return true if the simulation succeeds, and false if it times out and fails
      */
-    void setWorld(const World& new_world);
+    bool runSimulation(World world, const Duration& timeout);
 
    private:
-    /**
-     * Updates the simulation.
-     *
-     * This will advance the simulation by "simulation_time_step",
-     * "num_steps_per_primitive_update" times. We take several smaller steps if necessary
-     * to maintain the accuracy of the simulation. This will update the state of the World
-     * in the backend and send the updated values to all observers.
-     */
-    void updateSimulation();
-
-    /**
-     * Sends the current state of the world in the simulation to all Observers
-     */
-    void sendWorldToObservers();
-
     void onValueReceived(ConstPrimitiveVectorPtr primitives) override;
 
-    Duration simulation_time_step;
-    unsigned int num_steps_per_primitive_update;
-
-    ConstPrimitiveVectorPtr most_recently_received_primitives;
-    std::mutex most_recently_received_primitives_mutex;
+    const Duration physics_time_step;
+    const Duration world_time_increment;
+    SimulationSpeed simulation_speed_mode;
 
-    // The current state of the world in the simulation
-    World world;
-    std::mutex world_mutex;
+    std::promise<ConstPrimitiveVectorPtr> primitive_promise;
```

Comment:
Can we document how we're using promises and why they're needed?

Reply:
Will do once they're properly used. Should be one of the next PRs